File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        n, m, visited, area = len(grid), len(grid[0]), set(), 0
        indices = [(x,y) for x in range(n) for y in range(m)]
        self.ones = 0
        def dfs(row, col):
            self.ones += 1
            visited.add((row,col))
            
            for dirs in DIR:
                dx, dy = row + dirs[0], col + dirs[1]
                if (dx,dy) not in visited and inbound(dx,dy) and grid[dx][dy]:
                    dfs(dx,dy)
        
        DIR = [[0,1],[1,0],[-1,0],[0,-1]]
        inbound = lambda x, y: 0 <= x < n and 0 <= y < m
        for (x,y) in indices:
            if (x,y) not in visited and grid[x][y]:
                dfs(x,y)
                area = max(area, self.ones)
                self.ones = 0
                
        return area

File: 695-max-area-of-island/695-max-area-of-island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ans = 0, n = grid.size(), m = grid[0].size();

        function<int(int, int)> dfs = [&](int r, int c) {
            if (r < 0 || r >= n || c < 0 || c >= m)
                return 0;
            if (grid[r][c] == 0)
                return 0;
            grid[r][c] = 0;
            ans++;
            dfs(r + 1, c);
            dfs(r - 1, c);
            dfs(r, c + 1);
            dfs(r, c - 1);
            return ans;
        };
        
        int maxArea = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    ans = 0;
                    maxArea = max(maxArea, dfs(i, j));
                }
            }
        }
        return maxArea;
    }
};